fix: accept globs and array of globs for model option in config

darahayes opened the pull request:
fixes: https://github.com/aerogear/graphback/issues/977

For backwards compatibility specifying a dir like `./model` will still work as it did before, but it also means you can supply globs or arrays. For example, the following are both valid now.

```yaml
extensions:
  # Graphback configuration
  graphback:
    ##  Input schema
    model: ./model/**/*.graphql
```

```yaml
extensions:
  # Graphback configuration
  graphback:
    ##  Input schema
    model:
      - ./model/comment.graphql
     - ./model/note.graphql
```

There is also a custom [`loadSchema` function in graphql-serve](https://github.com/aerogear/graphback/blob/master/packages/graphql-serve/src/loadSchema.ts#L12-L24) which I am more than happy to update too if you think it's worth it

craicoverflow commented on packages/graphback-cli/src/components/generate.ts:
```diff
@@ -49,15 +50,37 @@ export const generateUsingPlugins = async (cliFlags: CliFlags) => {
+ */
+function getModelPath(baseDir: string, model: string | Array<string>): string | Array<string> {
+  if (typeof model === 'string' && existsSync(model) && lstatSync(model).isDirectory()) {
+    return join(baseDir, model, '/**/*.graphql')
```
Should we assume that

 **(a)** a basic `./model` directory means search `./model` and all of its subdirectories, or 
**(b)** should specifying `./model` mean only look in the `model` folder itself?

From a declarative point and just looking at the config without referring to the documentation it should be **(b)**.

This may break compatibility but if is wrong it's better to do this now before v1. What do you think?

I found this issue when trying out an example app on GitHub. The folder structure caused Graphback's generation to pick up `datamodel.graphql` and `generatedSchema.graphql` getting picked up as the data model because `./src` was the only model path that worked, causing some information to be lost from the final schema.

```
- src
  - datamodel.graphql
  - schema
      - generatedSchema.graphql
```

darahayes replied:
I would argue that the particular case you describe is a misconfiguration on the user's part but at the same time, I'm totally happy to go with option **b**. While it technically breaks compatibility, None of the current templates place things in subdirectories.

wtrocki replied:
If we phrase `./model` as corner case and backwards compatibility mode we should not care much as users will specify exactly what they looking for.

craicoverflow replied:
Let's go with **a** then, which is what you have :) Just docs updates are needed now.

craicoverflow commented on packages/graphback-cli/src/components/generate.ts:
```diff
@@ -49,15 +50,37 @@ export const generateUsingPlugins = async (cliFlags: CliFlags) => {
+ */
+function getModelPath(baseDir: string, model: string | Array<string>): string | Array<string> {
+  if (typeof model === 'string' && existsSync(model) && lstatSync(model).isDirectory()) {
+    return join(baseDir, model, '/**/*.graphql')
```
Thanks to graphql-config schema loaders, we should be able to use multiple input forms. This would mean that using `./model` will always assume `.graphql` file. This is absolutely fine - I think our documentation and templates should focus on specifying glob format so that this clear.

wtrocki commented:
To not go into a discussion with details about I think that proposed changes are removing couple really troubling issues and will help us to maintain backward compatibility and better flexibility for model creation.

wtrocki commented on templates/ts-apollo-runtime-backend/.graphqlrc.yml:
```diff
@@ -4,7 +4,7 @@ extensions:
   graphback:
     ##  Input schema
-    model: ./model
+    model: ./model/**/*.graphql
```
Needs doc update

darahayes commented on packages/graphback-cli/src/components/generate.ts:
```diff
@@ -35,9 +36,23 @@ export const generateUsingPlugins = async (cliFlags: CliFlags) => {
   const runGeneration = async () => {
     try {
-      const schemaDocument = await project.loadSchema(join(config.dirpath, graphbackConfig.model, '/**/*.graphql'));
+      const schemaDocument = await project.loadSchema(model);
```
I noticed I didn't include this in the first draft of the PR :)

darahayes commented on docs/intro/config.md:
```diff
@@ -33,8 +33,9 @@ documents: ./client/src/graphql/**/*.graphql
   graphback:
-    ##  Location to file that is contains GraphQL schema that will be processed by Graphback
-    model: ./model
+    ##  Path to file that contains GraphQL schema that will be processed by Graphback
+    ## Can be an array of paths too
+    model: ./model/**/*.graphql
```
@craicoverflow is this good enough or should I add some actual sentence to the doc explaining it?

wtrocki replied:
It is self explanatory for me

craicoverflow replied:
I think it is self explanatory :)

darahayes commented:
@craicoverflow thanks for the review, I've updated it now and I think it's a bit cleaner. I've also updated the docs and other places where we were using `./model` as a value

craicoverflow commented:
@darahayes I have just noticed that the runtime template fails to run because it loads the model the old way:

https://github.com/aerogear/graphback/blob/4a10783e73dd1d30522289cb8831ebbcdc712226/templates/ts-apollo-runtime-backend/src/loadSchema.ts#L12-L24

This can probably be quick fixed for now - I was in the process I removing the custom `loadSchema` from the template [here](https://github.com/aerogear/graphback/pull/1069/files#diff-132b30b3d9176017b87eba1eb7895816R25).

The model should not be loaded in the runtime template at all, so this can probably get resolved later in https://github.com/aerogear/graphback/issues/1071
